### PR TITLE
Added indexing of bag part content items and fixed simple query string search.

### PR DESCRIPTION
**Pull request by hswami**
Fixed issue #2431: Content items in a bag part are not indexed.
Fixed issue #2437: Search type "simple_query_string" throws an exception.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        public interface IDocumentIndexEntry
+        {
+            string Name { get; }
+            object Value { get; }
+            Types Type { get; }
+            DocumentIndexOptions Options { get; }
+        }
+
+        private class DocumentIndexEntry : IDocumentIndexEntry
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs]: Not sure there is a need to an interface. It's just a data object. We don't even do that anymore for services if there is no need to be able to replace them, or test them.

**Comment by sebastienros** [comment]
I assume you will still follow up with the comments I added in the other PR?

File: src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs (hunk 1 above)
Comment (hswami) [reply src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs]: OK good point ... I'll remove the interface and revert the abstraction.

**Comment by hswami** [comment]
>           I assume you will still follow up with the comments I added in the other PR?

Yes I will follow up with the comments from the other PR.

File: src/OrchardCore/OrchardCore.Indexing.Abstractions/ContentPartIndexHandler.cs [hunk 2]
```diff
@@ -10,18 +10,18 @@ namespace OrchardCore.Indexing
     /// </summary>
     public abstract class ContentPartIndexHandler<TPart> : IContentPartIndexHandler where TPart : ContentPart
     {
-        Task IContentPartIndexHandler.BuildIndexAsync(ContentPart contentPart, ContentTypePartDefinition typePartDefinition, BuildIndexContext context, ContentIndexSettings settings)
```
Comment (hswami) [inline src/OrchardCore/OrchardCore.Indexing.Abstractions/ContentPartIndexHandler.cs]: @sebastienros asked: Why are you adding async. It's better not to use the keyword when not necessary. Like in this case where the only statement to await is returned.

File: src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            if (bagPart.ContentItems.Count != 0)
+            {
+                // Lazy resolution to prevent cyclic dependency
+                var contentItemIndexHandlers = _serviceProvider.GetServices<IContentItemIndexHandler>();
+
+                foreach (var contentItemIndexHandler in contentItemIndexHandlers)
+                {
+                    foreach (var contentItem in bagPart.ContentItems)
+                    {
+                        var buildIndexContext = new BuildIndexContext(context.DocumentIndex, contentItem, contentItem.ContentType);
```
Comment (hswami) [inline src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs]: @sebastienros asked: When a bag part content item is indexed, shouldn't the keys be prefixed by the container keys?

File: src/OrchardCore/OrchardCore.Lucene.Core/QueryProviders/SimpleQueryStringQueryProvider.cs [hunk 4]
```diff
@@ -14,9 +15,19 @@ public Query CreateQuery(ILuceneQueryService builder, LuceneQueryContext context
             }
 
             var queryString = query["query"]?.Value<string>();
-            var fields = query["fields"]?.Value<string>() ?? "";
-            var defaultOperator = query["default_operator"]?.Value<string>() ?? "and";
```
Comment (hswami) [inline src/OrchardCore/OrchardCore.Lucene.Core/QueryProviders/SimpleQueryStringQueryProvider.cs]: @sebastienros asked: I would prefer not to change the default. I assume that the more terms you add, the less results you get. Unless the Elastic search states the opposite.

File: src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs (hunk 1 above)
Comment (hswami) [reply src/OrchardCore/OrchardCore.Indexing.Abstractions/DocumentIndex.cs]: Good point. I will revert the interface. That should revert quite a bit of code.

File: src/OrchardCore/OrchardCore.Lucene.Core/QueryProviders/SimpleQueryStringQueryProvider.cs (hunk 4 above)
Comment (hswami) [reply src/OrchardCore/OrchardCore.Lucene.Core/QueryProviders/SimpleQueryStringQueryProvider.cs]: Elastic Search does state the opposite. According to https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html, the default value for default_operator is OR.

File: src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs (hunk 3 above)
Comment (hswami) [reply src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs]: Good point. I will add the prefix using $"{context.Key}.{contentItem.ContentType}".

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs]: BagParts can be named, so i think it should use the part name. Please try the case when two BapParts with difference names are added to ensure it works fine -> we get the fields indexes under different prefixes.

Comment (hswami) [reply src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs]: Let me try that.

Comment (hswami) [reply src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs]: What about the aspects and default fields of a content item (AspectsContentIndexHandler and DefaultContentIndexHandler) ... currently they are all keyed off the values in IndexingConstants. Shouldn't these respect the context's key as well?

Comment (hswami) [reply src/OrchardCore.Modules/OrchardCore.Flows/Indexing/BagPartIndexHandler.cs]: ContentPartIndexHandler<TPart> doesn't honor this either. It uses the part definition name as the key with no prefix. Neither does ContentPickerFieldIndexHandler - is uses the context's key as-is. ContentFieldIndexHandler<TField> handles this only partially - it just uses the part definition name as the prefix followed by the field definition name.

File: src/OrchardCore/OrchardCore/Environment/Shell/ShellHost.cs [hunk 5]
```diff
@@ -20,7 +20,7 @@ namespace OrchardCore.Environment.Shell
     /// tenant is removed, but are necessary to match an incoming request, even if they are not initialized.
     /// Each <see cref="ShellContext"/> is activated (its service provider is built) on the first request.
     /// </summary>
-    public class ShellHost : IShellHost, IShellDescriptorManagerEventHandler
+    public class ShellHost : IShellHost, IShellDescriptorManagerEventHandler, IDisposable
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore/Environment/Shell/ShellHost.cs]: Isn't that from a different PR that has already been merged?

**Comment by hswami** [comment]
I'll just fix this branch ... not a new PR.

**Comment by hswami** [comment]
With the changes to ContentPartIndexHandler and ContentFieldIndexHandler, and no changes made to AspectsContentIndexHandler and DefaultContentIndexHandler, the field names in the index now look like this:

Content.BodyAspect.Body
Content.ContentItem.Author
Content.ContentItem.ContentItemId
Content.ContentItem.ContentItemVersionId
Content.ContentItem.ContentType
Content.ContentItem.CreatedUtc
Content.ContentItem.DisplayText
Content.ContentItem.DisplayText.Analyzed
Content.ContentItem.Latest
Content.ContentItem.ModifiedUtc
Content.ContentItem.Owner
Content.ContentItem.Published
Content.ContentItem.PublishedUtc
ContentItemId
LandingPage.About.Milestone.HtmlBodyPart
LandingPage.About.Milestone.Milestone.Date
LandingPage.About.Milestone.TitlePart
LandingPage.AutoroutePart
LandingPage.Clients.Client.TitlePart
LandingPage.Portfolio.Project.HtmlBodyPart
LandingPage.Portfolio.Project.Project.Category
LandingPage.Portfolio.Project.TitlePart
LandingPage.Services.Service.HtmlBodyPart
LandingPage.Services.Service.TitlePart
LandingPage.Team.TeamMember.TeamMember.Occupation
LandingPage.Team.TeamMember.TitlePart
LandingPage.TitlePart

Let me know if this is what you were expecting.

**Pull request merged**